vaadin-grid-flow-parent/vaadin-grid-flow/src/main/resources/META-INF/resources/frontend/gridConnector.js

```diff
@@ -716,23 +716,62 @@ import { ItemCache } from '@vaadin/vaadin-grid/src/vaadin-grid-data-provider-mix
         if (index % grid.pageSize != 0) {
           throw 'Got cleared data for index ' + index + ' which is not aligned with the page size of ' + grid.pageSize;
         }
+        // Generates an array of integers from 0 to size-1
+        const generateRange = size => [...Array(size).keys()];
 
-        let firstPage = Math.floor(index / grid.pageSize);
-        let updatedPageCount = Math.ceil(length / grid.pageSize);
+        function findDescendants(pkey,pages) {
+          const parentCache = cache[pkey];
 
-        for (let i = 0; i < updatedPageCount; i++) {
-          let page = firstPage + i;
-          let items = cache[pkey][page];
-          grid.$connector.doDeselection(items.filter(item => selectedKeys[item.key]));
-          delete cache[pkey][page];
-          const updatedItems = updateGridCache(page, parentKey);
-          if (updatedItems) {
-            itemsUpdated(updatedItems);
+          if(!parentCache) return [];
+
+          if(!pages) {
+            // When calling recursively, use all available pages
+            pages = generateRange(parentCache.length);
           }
-          updateGridItemsInDomBasedOnCache(items);
+
+          return pages.map(currentPage => {
+            const items = parentCache[currentPage] || [];
+            return items.concat(items.flatMap(item => findDescendants(item.key)));
+          }).reduce((current, value) => current.concat(value));
         }
+
+        const updatedPageCount = Math.ceil(length / grid.pageSize);
+        const firstPage = Math.floor(index / grid.pageSize);
+        const pages = generateRange(updatedPageCount).map(i => firstPage + i);
+
+        // Uses cache, so needs to run before deleting
+        const affectedItems = findDescendants(pkey,pages);
+
+        pages.forEach(page => {
+          delete cache[pkey][page];
+        });
+
+        // Use grid.clearCache if clearing the root cache.
+        if(!parentKey) {
+          grid.clearCache();
```

Comment:
Removed

Reply:
You still need to cover the root level case (other tests break because without it). The root level cache is just a regular item cache so you can use the same logic (clearSubcache) for clearing it.

Reply:
Done